### PR TITLE
Add DSA signature verification (OpenSSL based)

### PR DESCRIPTION
**Pull request by Youw**
Compatible with Sparkle method of DSA signing and verification.

Uses OpenSSL and adds external dependency: `libeay32.dll` (but NOT `ssleay32.dll`).

Theoretically can be reimplemented, to use WinAPI only.

Reffers to #23.

### CODE REVIEW

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately there were the following style issues with your Pull Request:

 * Commit: 458695052883e3bce413c25ff5e17b093261e924
  * Your subject line is longer than 50 characters

 * Commit: 6119e2e2df6395a6790b5759d8f68339d7b1f388
  * Your subject line is longer than 50 characters

Please follow [git commit message guidelines](http://chris.beams.io/posts/git-commit/) (you can use `git push -f` to update this PR).

---
This message was auto-generated by https://gitcop.com

**Comment by vslavik** [comment]
Thank you *a lot!*

Just two quick question before I actually look at the code:

> Compatible with Sparkle method of DSA signing and verification.

To confirm, this means that a signature produced by Sparkle's macOS tools will be accepted, right?

> Uses OpenSSL and adds external dependency: libeay32.dll (but NOT ssleay32.dll).

External *DLL* is unfortunate (I misunderstood this bit when we talked about adding an external dependency), but I think I have a fix for that: I use a VC++ project for statically-built libcrypto in something else, so I'll follow this PR with changes to use that and return into single-DLL packaging. Might be worth considering BoringSSL ([vcxproj](https://chromium.googlesource.com/external/github.com/grpc/grpc/+/chromium-deps/2016-07-19/vsprojects/vcxproj/boringssl)) or LibreSSL for that too, but again: don't worry about that part.

Did you use 1.0 or 1.1 series of OpenSSL?

**Comment by Youw** [comment]
> this means that a signature produced by Sparkle's macOS tools will be accepted

Right. And vice versa.

> Did you use 1.0 or 1.1 series of OpenSSL?

Good catch. I probably should have mention that in README.
I've used OpenSSL 1.0.2.

File: src/signatureverifier.cpp [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+
+#include "signatureverifier.h"
+
+#include <stdexcept>
+
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <Wincrypt.h>
+
+#pragma comment(lib, "crypt32.lib")
```
Comment (Youw) [inline src/signatureverifier.cpp]: Missed this - it is for Base64ToBin.
Does WinSparkle support MinGW/Cygwin builds?

**Comment by Youw** [comment]
BTW: let me know, if you require any additional assistance from me.

**Comment by Youw** [comment]
I've managed to port signature verification to bare WinAPI (so no `libeay32.dll`), but there is another problem: [CryptoAPI does not support DSA keys larger that 1024](https://msdn.microsoft.com/en-us/library/windows/desktop/bb931357(v=vs.85).aspx).

I have perfectly working code in [my repo](https://github.com/Youw/winsparkle/tree/win_crypto_dsa_verify), but it can use only DSA keys 1024 bits large. It partially breaks compatibility with Sparkle (and potentially less secure).

There is also a new API from Microsoft - [CNG](https://msdn.microsoft.com/en-us/library/windows/desktop/aa376210(v=vs.85).aspx), but it is available only starting from Windows Vista and DSA keys larger than 1024 supports only [starting from Window 8](https://msdn.microsoft.com/en-us/library/windows/desktop/aa375534(v=vs.85).aspx).

So, the only way to be fully compatible with Sparkle (and FIPS compliant) is to use OpenSSL on Windows.

Alternatives:
1) Use 1024 bit key - [security is no longer guaranteed with 1024 bit long keys](https://stackoverflow.com/a/16935946/3697939);
2) Switch to 4096 RSA signatures - secure, but fully breaks compatibility with Sparkle project.

@vslavik, thoughts?

File: src/signatureverifier.cpp (hunk 1 above)
Comment (vslavik) [reply src/signatureverifier.cpp]: Somewhat — there's the cmake build system that is meant to work with it, but it's not maintained by me. In any case, that's a trivial thing to `#ifdef`, so don't worry about that.

**Comment by vslavik** [comment]
First of all, sorry for not reviewing this before Xmas as I hoped to — and thanks a lot for your patience and continued work. I'll do my best to merge this this week.

Re DSA: wow, I had no idea of this mess.

So yes, this is an argument for sticking to OpenSSL. Another alternative is

3. Have both implementations and document that if you support anything other than Windows 8, you have to stick to 1024 keys — an acceptable tradeoff for supporting old OSes. If Windows 7 was the cutoff, I think I would be OK with that, but Windows 8 is still a bit high requirement. 

Finally, Sparkle has an open ticket about migrating off DSA at https://github.com/sparkle-project/Sparkle/issues/1037. That opens another option:

4. Agree with Sparkle guys on what to use and how and implement only that, i.e. being compatible with latest/future Sparkle.

On the other hand: you have already done all the hard work to support DSA via OpenSSL. Would be shame to waste that. On the *third* hand, the moment we release DSA support, we're stuck with supporting it forever.

File: src/signatureverifier.cpp (hunk 1 above)
Comment (Youw) [reply src/signatureverifier.cpp]: I just don't feel right, if I keep it as is. [9d811dd](https://github.com/Youw/winsparkle/commit/9d811dd2840ce6795d17ca870514a01d523834a2) =)

Comment (Youw) [reply src/signatureverifier.cpp]: BTW: cmake build is broken after one of the expat updates

Comment (vslavik) [reply src/signatureverifier.cpp]: @Youw Pro-tip: when you find yourself writing "BTW" in an issue tracker, stop right there — you're being off-topic. If you have a *completely unrelated* issue, file an unrelated issue in the tracker. In the case of cmake build, make that a PR (see the [explanation](https://github.com/vslavik/winsparkle/tree/master/cmake) why).

Comment (vslavik) [reply src/signatureverifier.cpp]: @Youw I'm sorry, let me rephrase: since the cmake build is unsupported, you don't have to worry about it as part of this PR.

Comment (Youw) [reply src/signatureverifier.cpp]: Sure, totally agree.

File: src/settings.h [hunk 2]
```diff
@@ -111,6 +111,15 @@ class Settings
         return ms_registryPath;
     }
 
+    /// Return DSA public key to verify update file signature
+    static std::string GetDSAPubKeyPem()
+    {
+        CriticalSectionLocker lock(ms_csVars);
+        if (ms_DSAPubKey.empty())
+            SetDSAPubKeyPem(GetCustomResource("DSAPub", "DSAPEM"));
```
Comment (vslavik) [inline src/settings.h]: I know (now…took me writing first version of this comment to realize it) that this works, because Windows critical sections are re-entrant, but I think it's not very intuitive — I expect this to deadlock because `SetDSAPubKeyPem` also holds `ms_csVars`.

I would prefer to do this in the same way as `GetAppcastURL()` above, i.e. just assign into `ms_DSAPubKey`. This  would have consequences for other code, of course — I'll comment on them in their respective places.

File: src/signatureverifier.cpp [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+        }
+
+    }; // DSAWrap
+
+    DSAPub m_dsaPub;
+    DSAPub &PubKey()
+    {
+        if (static_cast<DSA*>(m_dsaPub) == NULL)
+        {
+            std::string pem = Settings::GetDSAPubKeyPem(); // side effect - it will call SignatureVerifier::SetDSAPubKeyPem
```
Comment (vslavik) [inline src/signatureverifier.cpp]: Such side-effects should be, generally speaking, avoided.

File: src/updatedownloader.cpp [hunk 4]
```diff
@@ -177,6 +179,14 @@ void UpdateDownloader::Run()
       UpdateDownloadSink sink(*this, tmpdir);
       DownloadFile(m_appcast.DownloadURL, &sink, this);
       sink.Close();
+
+      if (!SignatureVerifier::DSASHA1SignatureValid(sink.GetFilePath(), m_appcast.DsaSignature))
+      {
+          // remove potentially corrupted file
+          _wremove(sink.GetFilePath().c_str());
```
Comment (vslavik) [inline src/updatedownloader.cpp]: Isn't it better to just call `CleanLeftovers()` (which does full cleanup) here?

File: src/updatedownloader.cpp [hunk 5]
```diff
@@ -177,6 +179,14 @@ void UpdateDownloader::Run()
       UpdateDownloadSink sink(*this, tmpdir);
       DownloadFile(m_appcast.DownloadURL, &sink, this);
       sink.Close();
+
+      if (!SignatureVerifier::DSASHA1SignatureValid(sink.GetFilePath(), m_appcast.DsaSignature))
```
Comment (vslavik) [inline src/updatedownloader.cpp]: This block should only be called if a public key is set. This is the promised comment on how I think setting the public key should be done better:

1. Only store it in `Settings`, don't do anything extra.
2. Here, obtain the key from `Settings:: GetDSAPubKeyPem()`. If not empty (but regardless of whether `m_appcast.DsaSignature` is empty or not, because that one is untrusted), call this block, otherwise don't. 
3. Pass the public key as argument to `DSASHA1SignatureValid`, or set it here, rather than doing that at some earlier point.

That would have the advantage that it would also work without OpenSSL DLLs if signatures are not used (which as far as I can tell is not the case with the current code due to setting the key possibly earlier?).

File: src/signatureverifier.cpp [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+
+        DSAPub &pubKey(PubKey());
+
+        const int code = DSA_verify(0, sha1, ARRAYSIZE(sha1), (const unsigned char*)signature.c_str(), signature.size(), pubKey);
+
+        if (code == -1) // OpenSSL error
+            throw std::runtime_error(ERR_error_string(ERR_get_error(), nullptr));
+
+        if (code != 1)
+            throw std::runtime_error("DSA Signature not match!");
```
Comment (vslavik) [inline src/signatureverifier.cpp]: Let's use just "Invalid signature".

File: src/updatedownloader.cpp (hunk 4 above)
Comment (Youw) [reply src/updatedownloader.cpp]: Sure - haven't seen it.

File: src/updatedownloader.cpp (hunk 5 above)
Comment (Youw) [reply src/updatedownloader.cpp]: OK, I got the idea (basically we keep backward compatibility)
> don't do anything extra

Maybe I'll remove key caching, but replace it with key verification
It will also address `side effect` issue

**Comment by Youw** [comment]
Thanks for comments. I'll update this PR within couple of days.
So I guess you're not going to add static libcrypto, and leave it as is?
> Also, could you send me (my email is in my github info) the private key you used so that I can properly sign the example download at winsparkle.org?

I thought you might want to generate your own keypair, and replace this public key instead.
But I'll send it to you anyway.

**Comment by vslavik** [comment]
> So I guess you're not going to add static libcrypto, and leave it as is?

I'll make the change to use static lib later, so that I don't interfere with your changes. 

> I thought you might want to generate your own keypair, and replace this public key instead.

I wanted to avoid unneeded commits...

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately there were the following style issues with your Pull Request:

 * Commit: 6cbf777f44c093ed3621ca9be4861c5518c7efae
  * Your subject line is longer than 50 characters

Please follow [git commit message guidelines](http://chris.beams.io/posts/git-commit/) (you can use `git push -f` to update this PR).

---
This message was auto-generated by https://gitcop.com

**Comment by vslavik** [comment]
I pushed static OpenSSL build to the branch.  The size difference of `WinSparkle.dll` is not as bad as I feared it would be:

* version from master is 1,387,520 bytes
* signature-verifying statically compiled version is 1,894,400 bytes

That's a ~36%, ~500 kB difference, I consider that acceptable. Ideally I'd investigate whether it could be made smaller with [libtomcrypt](https://github.com/libtom/libtomcrypt) (being my preference due to reputation, small footprint and public domain license), [Botan](https://botan.randombit.net) or [Crypto++](https://github.com/weidai11/cryptopp) to cut on dependency size, but 500 kB is fine for 0.6.0.

File: src/signatureverifier.cpp [hunk 7]
```diff
@@ -259,14 +203,6 @@ class TinySSL
 private:
     static HMODULE ms_libeay32;
```
Comment (Youw) [inline src/signatureverifier.cpp]: remove it too

File: src/signatureverifier.cpp [hunk 8]
```diff
@@ -329,14 +265,6 @@ class TinySSL
 
 HMODULE TinySSL::ms_libeay32 = NULL;
```
Comment (Youw) [inline src/signatureverifier.cpp]: remove

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately there were the following style issues with your Pull Request:

 * Commit: ea0a09554885d2b64ce41c4abaa6689bcdb92b13
  * Your subject line is longer than 50 characters

Please follow [git commit message guidelines](http://chris.beams.io/posts/git-commit/) (you can use `git push -f` to update this PR).

---
This message was auto-generated by https://gitcop.com

File: src/signatureverifier.cpp (hunk 8 above)
Comment (vslavik) [reply src/signatureverifier.cpp]: Thanks!

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately there were the following style issues with your Pull Request:

 * Commit: ea0a09554885d2b64ce41c4abaa6689bcdb92b13
  * Your subject line is longer than 50 characters

 * Commit: 92f526a4378ff8f1f13d4675c7ded5965830883b
  * Your commit message body contains a line that is longer than 72 characters

Please follow [git commit message guidelines](http://chris.beams.io/posts/git-commit/) (you can use `git push -f` to update this PR).

---
This message was auto-generated by https://gitcop.com

**Comment by vslavik** [comment]
I added one more thing: dedicated message for signature failure, as Sparkle does. It is important enough failure state to specifically alert the user to. I think we're good to go with this.

Thanks again for your work on this PR, it's much appreciated!

**Comment by Youw** [comment]
Nicely done!

I think it also makes sense to use `BadSignatureException` [here](https://github.com/vslavik/winsparkle/pull/157/commits/92f526a4378ff8f1f13d4675c7ded5965830883b#diff-58370b0e0e98f68e86ee09546e10b2eeL216), [here](https://github.com/vslavik/winsparkle/pull/157/commits/92f526a4378ff8f1f13d4675c7ded5965830883b#diff-58370b0e0e98f68e86ee09546e10b2eeL245) and [here](https://github.com/vslavik/winsparkle/pull/157/commits/92f526a4378ff8f1f13d4675c7ded5965830883b#diff-58370b0e0e98f68e86ee09546e10b2eeL281).

I guess everywhere in `signatureverifier.cpp`.

**Comment by vslavik** [comment]
> I think it also makes sense to use BadSignatureException here, here and here.

My reasoning for using it only in a few places was that it should be used when the code is (reasonably) certain the issue is a bad signature. If the problem is something else, such as invalid input, existing exception classes are more suitable (e.g. `invalid_argument`).

**Comment by Youw** [comment]
Good point.

I missed the fact that [this](https://github.com/vslavik/winsparkle/pull/157/commits/92f526a4378ff8f1f13d4675c7ded5965830883b#diff-58370b0e0e98f68e86ee09546e10b2eeR309) fixes all issues I was wanted to avoid.

But there is still one more issue worth considering: logging.
Maybe it worth:
```
    catch (BadSignatureException&)
    {
        throw;
    }
    catch (const std::exception &e)
    {
        throw BadSignatureException(e.what());
    }
    catch (...)
    {
        throw BadSignatureException("Unknown signature verification error");
    }
```

Right now, it `dsaSignature` xml key is broken, we're getting in logs:
```
Winsparkle:
```
And that's it.

**Comment by vslavik** [comment]
> But there is still one more issue worth considering: logging.

Done, thanks!

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately there were the following style issues with your Pull Request:

 * Commit: ea0a09554885d2b64ce41c4abaa6689bcdb92b13
  * Your subject line is longer than 50 characters

 * Commit: 92f526a4378ff8f1f13d4675c7ded5965830883b
  * Your commit message body contains a line that is longer than 72 characters

Please follow [git commit message guidelines](http://chris.beams.io/posts/git-commit/) (you can use `git push -f` to update this PR).

---
This message was auto-generated by https://gitcop.com
